feat(linter): react/exhaustive-deps

camc314 opened the pull request:
Firstly, a massive thanks to @alisnic for starting this (incredibly complicated) lint rule in https://github.com/oxc-project/oxc/pull/2637 !

closes  #2637
relates to #2174

camc314 commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3321 @@
+    match expr {
+        Expression::Identifier(ident) => {
+            // TODO: @overlookmotel how can i avoid the unsafe???
+            let ident = unsafe {
+                std::mem::transmute::<&IdentifierReference<'_>, &'a IdentifierReference<'a>>(ident)
+            };
```
any ideas on how to get around this?

camc314 replied:
@overlookmotel / @Boshen  ? 🙂

Boshen replied:
Store the name / span / id directly inside `Dependency`?

camc314 replied:
good shout 🙂

camc314 commented:
Rough first draft but most of the way their. would appreciate some early reviews as this was a massive amount of work and i don't want to go too far 🙂

Boshen commented:
> Rough first draft but most of the way their. would appreciate some early reviews as this was a massive amount of work and i don't want to go too far 🙂

This is amazing! I would run it through oxlint ecosystem ci and see what happens.

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+}
+
+fn unknown_dependencies_diagnostic(hook_name: &str, span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn(format!(
+      "React Hook {hook_name} received a function whose dependencies are unknown. Pass an inline function instead."
+    ))
+    .with_label(span)
```
Should this be help text?
```suggestion
    OxcDiagnostic::warn(format!(
      "React Hook {hook_name} received a function whose dependencies are unknown."
    ))
    .with_help("Pass an inline function instead.")
    .with_label(span)
```

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+}
+
+fn async_effect_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Effect callbacks are synchronous to prevent race conditions.")
+        .with_label(span)
```
```suggestion
    OxcDiagnostic::warn("Effect callbacks are synchronous to prevent race conditions.")
        .with_help("Remove `async` from the callback to make it synchronous.")
        .with_label(span)
```

camc314 replied:
the react help is a lot more complex here
i changed the help to `"Consider putting the asynchronous code inside a function and calling it from the effect."`

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+}
+
+fn literal_in_dependency_array_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("The literal is not a valid dependency because it never changes.")
+        .with_label(span)
```
```suggestion
    OxcDiagnostic::warn("The literal is not a valid dependency because it never changes.")
        .with_help("Remove the literal from the array.")
        .with_label(span)
```

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+}
+
+fn complex_expression_in_dependency_array_diagnostic(hook_name: &str, span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn(format!(
+        "React Hook {hook_name} has a complex expression in the dependency array. Extract it to a separate variable so it can be statically checked.",
+    ))
+    .with_label(span)
```
```suggestion
    OxcDiagnostic::warn(format!(
        "React Hook {hook_name} has a complex expression in the dependency array.",
    ))
    .with_help("Extract the expression to a separate variable so it can be statically checked.")
    .with_label(span)
```

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+}
+
+fn dependency_changes_on_every_render_diagnostic(hook_name: &str, span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn(format!(
+        "React Hook {hook_name} has a dependency array that changes every render."
+    ))
+    .with_label(span)
```
We should suggest memoizing I think, bonus points if we can point to the original definition of the variable with a span
```suggestion
    OxcDiagnostic::warn(format!(
        "React Hook {hook_name} has a dependency array that changes every render."
    ))
    .with_help("Try memoizing this variable with `useRef` or `useCallback`.")
    .with_label(span)
```

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+        "React Hook {hook_name} has an unnecessary dependency: {dep_name}."
+    ))
+    .with_label(span)
+    .with_help("Consider remove it from the dependency array. Outer scope values like aren't valid dependencies because mutating them doesn't re-render the component.")
```
```suggestion
    .with_help("Consider removing it from the dependency array. Outer scope values like aren't valid dependencies because mutating them doesn't re-render the component.")
```

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+fn get_node_name_without_react_namespace<'a, 'b>(expr: &'b Expression<'a>) -> Option<&'b Atom<'a>> {
+    match expr {
+        Expression::StaticMemberExpression(member) => {
+            if let Expression::Identifier(_ident) = &member.object {
```
```suggestion
            if let Expression::Identifier(_) = &member.object {
```

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+
+impl Dependency<'_> {
+    fn to_string(&self) -> String {
+        [vec![self.iref.name.clone()], self.chain.clone()].concat().join(".")
```
I would recommend avoiding creating a vec here, can we push to a string directly instead using `String.push` and `String.push_str`?

camc314 replied:
updated to use an iterator
```rs
std::iter::once(&self.iref.name)
    .chain(self.chain.iter())
    .map(|s| s.as_str())
    .join(".")
```

which should avoid the vec?

camchenry commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3454 @@
+    }
+}
+
+#[test]
+fn test() {
```
there are enough tests here that it might be a good idea to move it into a separate file? not sure what our standards here though are @Boshen

Boshen replied:
1 file is fine, makes it easy to search for things.

camchenry reviewed (commented):
This looks pretty good overall, I didn't thoroughly review all of the code (there is a lot!) but I took a look at the diagnostics and have a few help text suggestions. I think keeping this in nursery and continuing to iterate on it would be good

Boshen commented:
Note: once merged, manually add this rule to oxlint-ecosystem-ci's CI script. We then change this to correctness after running it for 2 - 4 weeks.

camchenry commented on crates/oxc_linter/src/service/runtime.rs:
```diff
@@ -206,6 +206,8 @@ impl Runtime {
         // The semantic model is not built at this stage.
         let semantic_builder = SemanticBuilder::new()
             .with_cfg(true)
+            // TODO: this is only used by exhaustive_deps, consider conditionally enabling?
+            .with_scope_tree_child_ids(true)
```
I gave it a shot, I think this works:

```rust
let should_build_child_ids = self.linter.rules.iter().any(|rule| matches!(rule.rule, RuleEnum::ExhaustiveDeps(_)));
```

```suggestion
            .with_scope_tree_child_ids(should_build_child_ids)
```

camc314 replied:
hmm let me get the benchmark working and look at the perf hit (if any)

camc314 replied:
<img width="1130" alt="Screenshot 2024-11-09 at 15 58 35" src="https://github.com/user-attachments/assets/794d48c5-ab73-4527-ab11-6a34797989e0">

shall we leave for now? look like there's no perf change


(i'm just about to put a comit to disable running this rule on non-`tsx/jsx` files to fix the checker.ts perf regression)

camchenry replied:
It might just be random, that doesn't seem like a big perf change (if any) to me so I think it's fine

camchenry reviewed (approved):
it's hard to review all of this code, but I have skimmed over most of it and not too much stuck out to me. I think this is ready to start being tested in nursery

camc314 commented:
@Boshen did it look ok in the ecosystem ci? https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/11755920122/job/32751524336

camc314 commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3467 @@
+        }
+    }
+
+    fn should_run(&self, ctx: &ContextHost) -> bool {
+        ctx.source_type().is_jsx()
+    }
```
can anyone think of a case where we would want this to run on non-jsx?

camc314 replied:
hmm i guess this is incorrect in the case where people are using `createElement` rather than jsx

camchenry replied:
Also lib files can be .ts files that have hooks in them with effects like https://github.com/streamich/react-use/

camc314 replied:
ah yes good thought

Boshen commented:
### Merge activity

* **Nov 11, 3:36 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Boshen commented:
Hit the merge button too quickly 😅 

In `oxlint-ecosystem-ci/repos`:

```
~/oxc/target/release/oxlint -A all -W exhaustive-deps
Finished in 5.5s on 117561 files with 1 rules using 8 threads.
Found 536 warnings and 23 errors.
```

Some false positives:

```
affine  canary ❯ ~/oxc/oxc/target/release/oxlint -A all -W exhaustive-deps

  ⚠ react_hooks(exhaustive-deps): React Hook useMemo has a missing dependency: 'T'
     ╭─[packages/frontend/core/src/components/page-list/virtualized-list.tsx:142:6]
 141 │     return items;
 142 │   }, [groupCollapsedState, groups]);
     ·      ─────────────────────────────
 143 │ };
     ╰────
  help: Either include it or remove the dependency array.
```

It's picking up time annotation

```
    const items: VirtuosoItem<T>[] = [];
```

Boshen commented:
Do you also need to include function calls?

```
  ⚠ react_hooks(exhaustive-deps): React Hook useEffect has a missing dependency: 'makeMappingRequest'
    ╭─[kibana/x-pack/plugins/enterprise_search/public/applications/enterprise_search_content/components/search_index/index_error.tsx:83:6]
 82 │     makeMappingRequest({ indexName });
 83 │   }, [indexName]);
    ·      ───────────
 84 │
    ╰────
  help: Either include it or remove the dependency array.
```

Boshen commented:
```
  ⚠ react_hooks(exhaustive-deps): React Hook useCallback has unnecessary dependency: dataVisualizerListState.sortField
     ╭─[kibana/x-pack/plugins/data_visualizer/public/application/index_data_visualizer/hooks/use_field_stats.ts:307:5]
 306 │     dataVisualizerListState.sortDirection,
 307 │     dataVisualizerListState.sortField,
     ·     ───────────────────────
 308 │     samplingProbability,
     ╰────
  help: Either include it or remove the dependency array.
```

The span is not pointing at the member expression.

Boshen commented:
This looks pretty weird

```
  ⚠ react_hooks(exhaustive-deps): React Hook useEffect has missing dependencies: 'searchParams.get', 't', and 'form.setValue'
    ╭─[cal/packages/features/eventtypes/components/DuplicateDialog.tsx:67:6]
 66 │     }
 67 │   }, [searchParams?.get("dialog")]);
    ·      ─────────────────────────────
 68 │
    ╰────
  help: Either include it or remove the dependency array.
```

Boshen commented:
kibana has the most violations, you may want to test it during developement.

camc314 commented:
> This looks pretty weird
> 
> ```
>   ⚠ react_hooks(exhaustive-deps): React Hook useEffect has missing dependencies: 'searchParams.get', 't', and 'form.setValue'
>     ╭─[cal/packages/features/eventtypes/components/DuplicateDialog.tsx:67:6]
>  66 │     }
>  67 │   }, [searchParams?.get("dialog")]);
>     ·      ─────────────────────────────
>  68 │
>     ╰────
>   help: Either include it or remove the dependency array.
> ``

🙂 looks like we're following the same behaviour as eslint


<img width="1130" alt="Screenshot 2024-11-11 at 09 40 29" src="https://github.com/user-attachments/assets/4adbf3f5-369f-4000-978e-73f392736ebf">

camc314 commented on crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:
```diff
@@ -0,0 +1,3483 @@
+                return false;
+            };
+
+            if (init_name == "useState"
+                || init_name == "useReducer"
+                || init_name == "useTransition")
```
TODO: fix

https://github.com/oxc-project/oxc/pull/7151#issuecomment-2467554800

https://github.com/facebook/react/blob/1b0132c05acabae5aebd32c2cadddfb16bda70bc/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L265-L320

camc314 commented:
> s/search_index/index_error.tsx:83:6

looks like they've got it disabled? for this file https://github.com/elastic/kibana//blob/2466a172af66452bdd939dddc56506faba7ffb7a/.eslintrc.js#L1623

when i run eslint on this file, it reports (same as us)

camc314 commented on crates/oxc_linter/src/snapshots/exhaustive_deps.snap:
```diff
@@ -0,0 +1,2224 @@
+   ·               ─────
+ 5 │         }
+   ╰────
+  help: Either include it or remove the dependency array.
```
fixed https://github.com/oxc-project/oxc/pull/7151/commits/e9144c31b1061d6c35861c2c02c37f919b057fc8

camc314 commented:
@Boshen after https://github.com/oxc-project/oxc/pull/7151/commits/a4f7675ef509ade89ccbc524bf5f2a80bb5f187c i don't there are more false positives in affine at least

Boshen commented:
```
  ⚠ react_hooks(exhaustive-deps): React Hook useCallback received a function whose dependencies are unknown.
    ╭─[kibana/src/plugins/console/public/application/hooks/use_save_current_text_object.ts:27:5]
 26 │       return useCallback(
 27 │ ╭─▶     throttle(
 28 │ │         (text: string) => {
 29 │ │           const { current: promise } = promiseChainRef;
 30 │ │           if (!currentTextObject) return;
 31 │ │           promise.finally(() =>
 32 │ │             objectStorageClient.text.update({ ...currentTextObject, text, updatedAt: Date.now() })
 33 │ │           );
 34 │ │         },
 35 │ │         WAIT_MS,
 36 │ │         { trailing: true }
 37 │ ╰─▶     ),
 38 │         [objectStorageClient, currentTextObject]
    ╰────
  help: Pass an inline function instead.
 ```
 
Interesting case, how should this be handled?

Boshen reviewed (approved):
Feel free to merge after a final self review.

camc314 commented:
> /use_save_current_text_object.ts:27:5



> ```
>   ⚠ react_hooks(exhaustive-deps): React Hook useCallback received a function whose dependencies are unknown.
>     ╭─[kibana/src/plugins/console/public/application/hooks/use_save_current_text_object.ts:27:5]
>  26 │       return useCallback(
>  27 │ ╭─▶     throttle(
>  28 │ │         (text: string) => {
>  29 │ │           const { current: promise } = promiseChainRef;
>  30 │ │           if (!currentTextObject) return;
>  31 │ │           promise.finally(() =>
>  32 │ │             objectStorageClient.text.update({ ...currentTextObject, text, updatedAt: Date.now() })
>  33 │ │           );
>  34 │ │         },
>  35 │ │         WAIT_MS,
>  36 │ │         { trailing: true }
>  37 │ ╰─▶     ),
>  38 │         [objectStorageClient, currentTextObject]
>     ╰────
>   help: Pass an inline function instead.
> ```
> 
> Interesting case, how should this be handled?

lol we do the same thing as react, only we're currently reporting on a different line so it wasn't being ignored


<img width="1137" alt="Screenshot 2024-11-12 at 08 54 34" src="https://github.com/user-attachments/assets/97b310e2-4385-472c-85f1-39fc3e60f267">

camc314 commented:
@Boshen i guess we should be reporting the same (or at least very similar spans to react? if ok with you, i'll make an issue to track bits that aren't exactly complete for this rule

Boshen commented:
> @Boshen i guess we should be reporting the same (or at least very similar spans to react? if ok with you, i'll make an issue to track bits that aren't exactly complete for this rule

Indeed, current span breaks already written eslint ignores.

camc314 committed:
feat(linter): react/exhaustive-deps (#7151)

Firstly, a massive thanks to @alisnic for starting this (incredibly complicated) lint rule in https://github.com/oxc-project/oxc/pull/2637 !

still a draft. current state:

3x false positives (all todo with refs)

3x false negatives (TS will catch these
13x false negatvies todo with refs
1x false negative TODO

closes  #2637
relates to #2174
